### PR TITLE
Added speed limits support, refactored events retrieval.

### PR DESCRIPTION
**Pull request by yunikkk**

### CODE REVIEW

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt [hunk 1]
```diff
@@ -158,11 +150,8 @@ object VisionManager : ARDataProvider {
     }
 
     // Event Listeners
-    private var visionEventsListener: VisionEventsListener? by Delegates.observable<VisionEventsListener?>(null) { _, _, new ->
-        if (::visionCore.isInitialized) {
-            visionCore.setVisionEventListener(new)
-        }
-    }
+    private var visionEventsListener: WeakReference<VisionEventsListener>? = null
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: I've made them all wear references to align approaches.

File: MapboxVision/src/main/java/com/mapbox/vision/visionevents/events/classification/SignClassification.kt [hunk 2]
```diff
@@ -15,57 +15,52 @@ import com.mapbox.vision.visionevents.events.detection.Detection
...
-
-            var index = 0
-            for (i in 0 until signValuesItemsSize) {
-
-                val signTypeIndex = signClassificationDataBuffer.signValueItems[index++].toInt()
-
-                val signType = if (signTypeIndex < 0 || signTypeIndex >= SignType.values().size) {
-                    if(BuildConfig.DEBUG) {
-                        throw IllegalArgumentException(" Wrong type index $signTypeIndex")
+    internal constructor(signClassificationDataBuffer: SignClassificationDataBuffer) : this(
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/visionevents/events/classification/SignClassification.kt]: Moved static creators to plain constructors, where it was possible and easy.

File: MapboxVision/src/main/java/com/mapbox/vision/corewrapper/update/jni/JNICoreUpdateManager.kt [hunk 3]
```diff
@@ -188,14 +188,54 @@ internal class JNICoreUpdateManager(
...
 
-    // Private methods
-
-    // Detections
     private fun initDetectionBuffer() {
         detectionDataBuffer = DetectionDataBuffer()
         coreWrapper.setDetectionDataBuffer(detectionDataBuffer!!)
     }
 
+    private val detectionsSource = object : Image.ImageSource {
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/corewrapper/update/jni/JNICoreUpdateManager.kt]: Created listeners once instead on every frame.

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt [hunk 4]
```diff
@@ -7,15 +7,12 @@ import com.mapbox.android.telemetry.MapboxTelemetry
...
+import com.mapbox.vision.corewrapper.update.RoadRestrictionsListener
 import com.mapbox.vision.corewrapper.update.VisionEventsListener
 import com.mapbox.vision.location.LocationEngine
 import com.mapbox.vision.location.LocationEngineListener
 import com.mapbox.vision.location.android.AndroidLocationEngineImpl
-import com.mapbox.vision.models.CameraParamsData
-import com.mapbox.vision.models.DeviceMotionData
-import com.mapbox.vision.models.FrameStatistics
-import com.mapbox.vision.models.GPSData
-import com.mapbox.vision.models.HeadingData
```
Comment (BenDenen) [inline MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: Should we collapse it?

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt [hunk 5]
```diff
@@ -31,16 +28,11 @@ import com.mapbox.vision.video.videosource.VideoSource
 import com.mapbox.vision.video.videosource.VideoSourceListener
 import com.mapbox.vision.video.videosource.camera.CameraVideoSourceImpl
 import com.mapbox.vision.view.VisualizationUpdateListener
-import com.mapbox.vision.visionevents.CalibrationProgress
-import com.mapbox.vision.visionevents.FrameSize
-import com.mapbox.vision.visionevents.LaneDepartureState
-import com.mapbox.vision.visionevents.ScreenCoordinate
-import com.mapbox.vision.visionevents.WorldCoordinate
```
Comment (BenDenen) [inline MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: the same

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt (hunk 1 above)
Comment (BenDenen) [reply MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: Ok. It is fine for me. But may be it is better to wrap listener to WeakReference at SDK level?

File: MapboxVision/src/main/java/com/mapbox/vision/corewrapper/jni/JNIVisionCoreImpl.kt [hunk 6]
```diff
@@ -9,18 +9,15 @@ import com.mapbox.vision.core.events.EventManager
...
 import com.mapbox.vision.models.CameraParamsData
 import com.mapbox.vision.models.DeviceMotionData
 import com.mapbox.vision.models.GPSData
 import com.mapbox.vision.models.HeadingData
 import com.mapbox.vision.models.route.NavigationRoute
-import com.mapbox.vision.performance.ModelPerformance
-import com.mapbox.vision.performance.ModelPerformanceConfig
-import com.mapbox.vision.performance.ModelPerformanceMode
-import com.mapbox.vision.performance.ModelPerformanceRate
-import com.mapbox.vision.performance.PerformanceManagerFactory
```
Comment (BenDenen) [inline MapboxVision/src/main/java/com/mapbox/vision/corewrapper/jni/JNIVisionCoreImpl.kt]: The same about collapsing

File: MapboxVision/src/main/java/com/mapbox/vision/corewrapper/update/jni/JNICoreUpdateManager.kt [hunk 7]
```diff
@@ -37,38 +38,34 @@ internal class JNICoreUpdateManager(
     private val mainThreadHandler = MainThreadHandler().also { it.start() }
     private val visualizationUpdateThreadHandler = WorkThreadHandler("VisualizationUpdateThread").also { it.start() }
 
-    // Detection Event
```
Comment (BenDenen) [inline MapboxVision/src/main/java/com/mapbox/vision/corewrapper/update/jni/JNICoreUpdateManager.kt]: why you remove it? It is helpful for code navigation

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt (hunk 4 above)
Comment (yunikkk) [reply MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: Discussed

**Pull request merged**